pkg/pgmodel/ingestor/ingestor.go

```diff
@@ -47,20 +49,91 @@ func NewPgxIngestorForTests(conn pgxconn.PgxConn) (*DBIngestor, error) {
 	return NewPgxIngestor(conn, c, s, &Cfg{})
 }
 
+const (
+	meta = iota
+	series
+)
+
 // Ingest transforms and ingests the timeseries data into Timescale database.
 // input:
 //     tts the []Timeseries to insert
 //     req the WriteRequest backing tts. It will be added to our WriteRequest
 //         pool when it is no longer needed.
-func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (uint64, error) {
+func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (numSamples uint64, numMetadata uint64, err error) {
 	var (
-		err       error
-		totalRows uint64
+		timeseries = r.Timeseries
+		metadata   = r.Metadata
+	)
+	// WriteRequests can contain pointers into the original buffer we deserialized
+	// them out of, and can be quite large in and of themselves. In order to prevent
+	// memory blowup, and to allow faster deserializing, we recycle the WriteRequest
+	// here, allowing it to be either garbage collected or reused for a new request.
+	// In order for this to work correctly, any data we wish to keep using (e.g.
+	// samples) must no longer be reachable from req.
+	release := func() { FinishWriteRequest(r) }
+	switch numTs, numMeta := len(timeseries), len(metadata); {
+	case numTs == 0 && numMeta == 0:
+		release()
+		return 0, 0, nil
+	case numTs > 0 && numMeta == 0:
+		// Write request contains only time-series.
+		n, err := ingestor.ingestTimeseries(timeseries, release)
+		return n, 0, err
+	case numMeta > 0 && numTs == 0:
+		// Write request contains only metadata.
+		n, err := ingestor.ingestMetadata(metadata, release)
+		return 0, n, err
+	default:
+		release = func() {}
```

Comment:
Might be a good idea to leave the request cleanup to samples ingestion in this case, since metadata data should not be that big most cases and it would clean up the logic .

Reply:
Sorry, I do not get the point. Passing the cleanup in release callback I think looks cleaner so I kept it, plus we need to release on metadata. Metadata can be large especially if Prometheus is scraping kube-state-metrics target, as that contains many metrics and usually takes time to scrape by Prometheus server.

Reply:
The recent commit push cleans up this part too.